### ISSUE TITLE
Proposal: Adopting an explicit version of ks-jenkins

### ISSUE THREAD

**Issue by LinuxSuRen** [kubesphere/ks-jenkins#33]
Currently, there are a few problems with the version of `ks-jenkins`. Please see also the following items:

* no way to find the relationship between `ks-jenkins` and KubeSphere
* having the same version between `ks-jenkins` and Jenkins core makes it hard to maintain a real version of `ks-jenkins`
  * For example, we always need to change the Jenkins core version if we want to bump version of `ks-jenkins`

# Proposal

Having combination version with KubeSphere and Jenkins core.  For example, `v3.1.1-2.249.1`.

/cc @kubesphere/sig-devops 
/kind proposal

**Comment by JohnNiang** [comment]
Thanks for your proposal. Combination version `v3.1.1-2.249.1` is good to identify, but this style is not good enough. Best practice is following [SemVer](https://semver.org/). I have an idea about version of `ks-jenkins` as follow:

```text
<version core> ::= <major: kubesphere version> "." <minor: jenkins version> "." <patch>

eg: v310.2249.1, v311.22491.1, v311.22491.2
```

But there has a problem that we have to follow KubeSphere release. Imagine that KubeSphere has released v3.1.1, v3.1.2 and v3.2.0. We have to follow up the release of each version, even if there is no update for `ks Jenkins`.

In summary, my suggestion is to follow the [SemVer](https://semver.org/) simply. There is no need to connect version between `ks-jenkins` and KubeSphere.

**Comment by LinuxSuRen** [comment]
Semantic Version is not a silver bullet. It is not suitable for all use cases. The important is that `ks-jenkins` is not an independent component.

For this case, `v310.2249.1`. `v310` is not a major version. The patch part is not a patch. Let's say, what do we need to do if there's a patch version that needs to be released. I don't think the version `v310.2249.2` is appropriate if the Jenkins core version has not changed.

Secondly, you can find other cases that using the combination version. For example, [golang:1.16.4-alpine3.13
](https://hub.docker.com/layers/golang/library/golang/1.16.4-alpine3.13/images/sha256-8aef644698ddcdbb64d7be44b278bb12c4a1117296e347e49cb2532f705bb732?context=explore).

* `jenkins/jenkins:2.277.4-lts-alpine`
* `jenkins/jenkins:2.295-jdk11`

What's your suggestion If the above version adopts a semantic Version? I guess `v2295.11.0` is not beautiful.

**Comment by JohnNiang** [comment]
If we use version like v3.1.0-2.249.1, where patch version of `ks-jenkins` should be placed?

BTW, the version like `jenkins/jenkins:2.277.4-lts-alpine` has its own major, minor and patch version, but `v3.1.0-2.249.1` not.

**Comment by LinuxSuRen** [comment]
> If we use version like v3.1.0-2.249.1, where patch version of ks-jenkins should be placed?

`v3.1.0` is the semantic version part. So the patch version here is `0`.

**Comment by JohnNiang** [comment]
I think this version naming is not helpful for future independent `DevOps` components. What's your opinion?

**Comment by LinuxSuRen** [comment]
> I think this version naming is not helpful for future independent `DevOps` components. What's your opinion?

I didn't get your point. Please explain it. Why it is not helpful?

**Comment by JohnNiang** [comment]
Currently, we can adopt this version naming for experimental use. Thanks for your proposal again.

**Comment by JohnNiang** [comment]
Due to proposal accepted, I'm going to close this proposal.
/close
